### PR TITLE
feat: evaluate image-to-text model (blip and vision-encoder-decoder)

### PR DESCRIPTION
**Pull request by zhenchaoni**
# Image-to-text export and accuracy evaluation (BLIP + VED)

## Summary

1. Add split export for **BLIP** (`blip`) and **vision-encoder-decoder** (`vision-encoder-decoder`) on the `image-to-text` task, producing two ONNX artifacts (encoder + decoder) with a traceable **static KV cache** in the decoder. End-to-end inference works through HF's `pipeline("image-to-text", ...)` on NPU fp16.

2. The VED export is **fully polymorphic over inner decoder** — TrOCR (BART-family with learned or sinusoidal positions), MBart (donut, nougat), BERT (manga-ocr), and GPT-2 (vit-gpt2) all flow through the same `VisionDecoderWrapper`. No per-family wrapper class, no inner-decoder registry.

3. Encoder polymorphism: ViT-style scalar `image_size` (TrOCR / vit-gpt2 / manga-ocr) and Swin-style `[H, W]` rectangular images with hierarchical downsampling (donut, nougat) both supported via a single `encoder_seq_length` calculator.

4. **New `WinMLImageToTextEvaluator`** covering both captioning (BLIP, vit-gpt2) and OCR (TrOCR, donut, nougat) on the same `image-to-text` pipeline surface. Reports **CER** (manual Levenshtein) and **CIDEr** (pycocoevalcap) every run; the e2e harness picks the relevant metric per model via `winml_metric_key`.

5. **E2E accuracy plumbing for fp16-only models.** Top-level `precision` field in `models_with_acc.json`, honored by `run_eval.py` (`entry.precision or _DEFAULT_PRECISION`); `text2text-generation` mapped to `TextDataset` for calibration; CER added to `METRIC_COMPARE_STRATEGY` (lower-is-better, 5%/10%); five entries registered with `precision: fp16` (BLIP + four TrOCR variants).

## Architecture — Export

```
WinMLDecoderWrapper                                    # base; three-step adapter
└─ alias past KV inputs  →  invoke HF decoder forward  →  capture present KV outputs

WinMLStaticCacheDecoderIOConfig                        # base IOConfig
├─ past_key_input_names                                # semantic accessors used
└─ past_value_input_names                              # by the wrapper's _make_cache

BlipDecoderWrapper / BlipDecoderIOConfig               # BLIP — text decoder + 3-D mask
                                                       # bypasses HF's causal-triangle
                                                       # reconstruction (NPU-unfriendly).

VisionDecoderWrapper / VisionDecoderIOConfig           # VED — generic, polymorphic over
└─ inspects model.forward signature for                # inner decoder model_type
   ``position_ids`` and side-channels                  # (TrOCR / MBart / BERT / GPT-2)
   ``cache_position`` to bypass HF's
   dynamic position-ids derivation.

PATCHING_SPECS                                         # class-targeted, additive
├─ TrOCRLearnedPositionalEmbedding.forward             # bypass torch.arange(past_kv_len, ...)
└─ TrOCRSinusoidalPositionalEmbedding.forward          # bypass cumsum + dynamic Cast/Add

WinMLBlipImageToText  (hf/blip.py)                     # inference composite — colocated with
WinMLVEDImageToText   (hf/vision_encoder_decoder.py)   # the export wrapper, matching the
                                                       # bart/marian/t5/mu2/qwen convention.
```

### How NPU-unfriendly ops are avoided

The HF decoder forward emits patterns the OpenVINO/QNN analyzer can't lower under static-cache export. Each of the following is short-circuited:

| Source op chain | Where | Fix |
|---|---|---|
| `torch.arange(past_kv_len, past_kv_len + seq_len)` (TrOCR learned position) | `TrOCRLearnedPositionalEmbedding.forward` | Patcher reads side-channel `position_id` and skips `arange`. |
| `cumsum(input_ids != pad)` + dynamic Cast/Add (TrOCR sinusoidal) | `TrOCRSinusoidalPositionalEmbedding.forward` | Patcher reads side-channel `position_id`, shifted by `padding_idx + 1`. |
| BERT/GPT-2 internal `position_ids = arange(past_kv_len, ...)` derivation | inner decoder forward | `_invoke_hf` signature-inspects and passes `position_ids = cache_position` explicitly. |
| `arange + repeat + LessOrEqual` causal-triangle reconstruction (BLIP) | `BlipTextModel.get_extended_attention_mask` | Pass a 3-D `[1, 1, max_cache_len]` mask to route through the "already-causal" branch. |
| HF `cache.get_seq_length()` returning the literal `0` (frozen at trace time) | TrOCR mask shape | `_make_cache` overrides `cache.get_seq_length` to return the dynamic `cache_position`. |

### Polymorphism details

- `_VedDecoderNormalizedConfig.encoder_seq_length` — handles both scalar ViT `(H/p)^2 + 1` and rectangular Swin `(H/p/2^(N-1)) * (W/p/2^(N-1))` based on `image_size` shape and presence of `depths`.
- `_VedDecoderNormalizedConfig.num_layers` — prefers `decoder.decoder_layers` (BART/MBart family) over Optimum's default `num_hidden_layers`, which for BART/MBart aliases the encoder side and is wrong for VED.
- `WinMLVEDImageToText` mirrors that override onto `config.decoder.num_hidden_layers` so HF's `StaticCache.__init__` sizes the cache layers correctly for donut/nougat (where `decoder_layers=4` but `num_hidden_layers=12`).

## Architecture — Evaluation

```
WinMLImageToTextEvaluator                              # registered for "image-to-text"
├─ schema_info()                                       # image input + reference text (str|list)
└─ compute()                                           # iterate samples, aggregate metrics
    └─ TextSimilarityMetric                            # streaming aggregator
       ├─ CER   via _levenshtein                       # manual, no extra deps
       │                                               # multi-ref: min edit distance per sample
       │                                               # aggregate = total_edits / total_ref_chars
       └─ CIDEr via pycocoevalcap.cider.cider.Cider    # TF-IDF n-gram consensus
                                                       # IDF computed across the eval set
```

- HF pipeline returns either `[{"generated_text": ...}]` or `{"generated_text": ...}`; the evaluator normalises both and tolerates the rare bare-string case.
- Samples missing image or reference are counted under `skipped`; pipeline exceptions are caught per sample and counted as skipped (does not abort the run).
- CER aggregation matches the OCR convention: ratio of total edit distance to total reference characters (not the mean of per-sample CER).
- CIDEr requires multi-sample IDF computation; `compute()` runs once at the end of the pass over the full prediction/reference pool.

## Files

| File | Change |
|---|---|
| `models/hf/decoder_wrapper.py` | **New.** `WinMLDecoderWrapper` base (three-step adapter) + `WinMLStaticCacheDecoderIOConfig` (semantic accessors for past-KV input names). |
| `models/hf/blip.py` | **Rewritten.** Split encoder + decoder wrappers; KV-cache decoder; 3-D mask bypass; explicit `position_ids` to avoid baked-in `past_kv_len=0`. Composite inference class `WinMLBlipImageToText` colocated. |
| `models/hf/vision_encoder_decoder.py` | **New.** `VisionEncoderWrapper` + `VisionDecoderWrapper` (generic, polymorphic on `model.decoder`); `VisionDecoderIOConfig`; `PATCHING_SPECS` for TrOCR learned + sinusoidal positional embeddings; `_VedDecoderNormalizedConfig` handling encoder/decoder field-name and shape variation. Composite inference class `WinMLVEDImageToText` colocated. |
| `models/winml/encoder_decoder.py`, `kv_cache.py` | (Reused as-is.) |
| `eval/image_to_text_evaluator.py` | **New.** `WinMLImageToTextEvaluator`. |
| `eval/metrics/text_similarity.py` | **New.** `TextSimilarityMetric` (CER + CIDEr) and `_levenshtein` helper. |
| `eval/evaluate.py` | Register `"image-to-text"` → `WinMLImageToTextEvaluator`. |
| `datasets/__init__.py` | Map `"text2text-generation"` → `TextDataset` for quantization calibration. |
| `scripts/e2e_eval/utils/registry.py` | Add `precision: str \| None` field to `ModelEntry`; load from registry JSON. |
| `scripts/e2e_eval/run_eval.py` | Pass `entry.precision or _DEFAULT_PRECISION` into the build. |
| `scripts/e2e_eval/utils/accuracy.py` | Add `"cer"` entry to `METRIC_COMPARE_STRATEGY` (lower-is-better, 5%/10%). |
| `scripts/e2e_eval/testsets/models_with_acc.json` | Five new image-to-text entries with `precision: fp16` (BLIP, trocr-base/large × handwritten/printed). |
| `scripts/e2e_eval/cache/baseline_cache.json` | Updated baseline entries from PyTorch-CPU runs. |
| `pyproject.toml` | Add `pycocoevalcap>=1.2`. |
| `tests/unit/export/test_blip_onnx_config.py` | Unit tests for BLIP encoder/decoder I/O specs and registry. |
| `tests/unit/export/test_vision_encoder_decoder_onnx_config.py` | Unit tests for VED I/O specs, registry, `num_layers` BART/non-BART dispatch, and `encoder_seq_length` scalar/HxW dispatch. |
| `tests/unit/eval/test_image_to_text_evaluator.py` | **New.** Schema, init, registry, compute (perfect / dict-shape / missing-data skip / pipeline-exception skip / custom columns / empty dataset). |
| `tests/unit/eval/test_text_similarity_metric.py` | **New.** Levenshtein edge cases, CER (perfect / single-sub / aggregate / multi-ref / garbage), CIDEr (perfect-positive / disjoint-zero), result shape. |

### CODE REVIEW

File: src/winml/modelkit/models/hf/decoder_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        """
+        cfg = self.onnx_config
+        # past_i_key/value layout: [batch, heads, max_cache_len, head_dim]
+        key_names = cfg.past_key_input_names
+        value_names = cfg.past_value_input_names
+        sample_k = inputs[key_names[0]]
+        decoder_config = self.config.get_text_config(decoder=True)
+        cache = self._CACHE_CLS(decoder_config, max_cache_len=sample_k.size(2))
+        cache.early_initialization(
+            batch_size=inputs[cfg.decoder_input_ids_name].size(0),
```
Comment (vortex-captain) [inline src/winml/modelkit/models/hf/decoder_wrapper.py]:
```suggestion
            batch_size=sample_k.size(0),
```

File: src/winml/modelkit/models/hf/decoder_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+from optimum.exporters.onnx import OnnxConfig
+
+from ..winml.kv_cache import WinMLStaticCache
+
+
+if TYPE_CHECKING:
+    from ..winml.kv_cache import WinMLCache
+
+
+class WinMLDecoderWrapper(nn.Module):
```
Comment (vortex-captain) [inline src/winml/modelkit/models/hf/decoder_wrapper.py]: abstract class?

File: src/winml/modelkit/models/hf/decoder_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    family is free to pick its own ONNX input names).  Override these
+    accessors only if the family uses non-default names — for example a
+    decoder-only LM exporter that prefers ``input_ids`` over
+    ``decoder_input_ids``.
+
+    Defaults match HuggingFace's encoder-decoder convention.
+    """
+
+    @property
+    def decoder_input_ids_name(self) -> str:
```
Comment (vortex-captain) [inline src/winml/modelkit/models/hf/decoder_wrapper.py]: could be removed?

File: src/winml/modelkit/models/hf/tocr.py [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+    HIDDEN_SIZE = "decoder.d_model"
+    NUM_LAYERS = "decoder.decoder_layers"
+    NUM_ATTENTION_HEADS = "decoder.decoder_attention_heads"
+    MAX_CACHE_LEN = "decoder.max_position_embeddings"
+
+    @property
+    def head_dim(self) -> int:
+        return self.hidden_size // self.num_attention_heads
+
+    @property
```
Comment (vortex-captain) [inline src/winml/modelkit/models/hf/tocr.py]: maybe `ENCODER_HIDDEN_SIZE = "encoder.hidden_size"` ? (simiarly for the other 2)

File: src/winml/modelkit/models/hf/tocr.py [hunk 5]
```diff
@@ -0,0 +1,301 @@
+# -------------------------------------------------------------------------
```
Comment (vortex-captain) [inline src/winml/modelkit/models/hf/tocr.py]: to double confirm the file and class names: `tocr` or `trocr`?

Comment (zhenchaoni) [reply src/winml/modelkit/models/hf/tocr.py]: It should be trocr. It is my typo, I will fix in the next commit.

File: src/winml/modelkit/models/hf/tocr.py [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+# =============================================================================
+# Internal: TrOCR static-shape patches for ONNX export
+# =============================================================================
+# TrOCRAttention reshapes Q/K/V to ``(bsz * num_heads, ...)`` and
+# TrOCRLearnedPositionalEmbedding builds positions via ``torch.arange`` —
+# both produce symbolic shapes that NPU compilers reject.  The patches below
+# replace both with static-shape equivalents only during ``torch.onnx.export``.
+
+
+def _patched_tocr_attention_forward(
```
Comment (vortex-captain) [inline src/winml/modelkit/models/hf/tocr.py]: reminder to confirm if https://github.com/microsoft/WinML-ModelKit/commit/35464830d3d6aed99e0b3a6343e4854e9b205e4c works and simplifies the code

File: src/winml/modelkit/models/hf/decoder_wrapper.py (hunk 2 above)
Comment (zhenchaoni) [reply src/winml/modelkit/models/hf/decoder_wrapper.py]: Yes. Resolve this comment by making this class abstract.

File: src/winml/modelkit/models/hf/tocr.py (hunk 4 above)
Comment (zhenchaoni) [reply src/winml/modelkit/models/hf/tocr.py]: Fixed

File: src/winml/modelkit/models/hf/tocr.py (hunk 6 above)
Comment (zhenchaoni) [reply src/winml/modelkit/models/hf/tocr.py]: Yes, your commit is a much cleaner way to patch the forward method for the Range op explicitly. It is also working from e2e.

File: src/winml/modelkit/models/hf/decoder_wrapper.py (hunk 3 above)
Comment (zhenchaoni) [reply src/winml/modelkit/models/hf/decoder_wrapper.py]: yes, removed

File: src/winml/modelkit/models/winml/image_to_text.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+"""WinML inference classes for image-to-text composite models.
```
Comment (vortex-captain) [inline src/winml/modelkit/models/winml/image_to_text.py]: nit: other composite models (for inference) are defined and registered in `hf/`, would it be easy to move `WinMLBlipImageToText` and `WinMLVEDImageToText` to `hf/`?

File: src/winml/modelkit/models/hf/vision_encoder_decoder.py [hunk 8]
```diff
@@ -2,16 +2,498 @@
...
 # =============================================================================
 # WinML Build Config
 # =============================================================================
+
 VISION_ENCODER_DECODER_CONFIG = WinMLBuildConfig(
     optim=WinMLOptimizationConfig(
+        gelu_fusion=True,
+        layer_norm_fusion=True,
+        matmul_add_fusion=True,
+        clip_constant_values=True,
```
Comment (vortex-captain) [inline src/winml/modelkit/models/hf/vision_encoder_decoder.py]: nit: seems the models we are working on here do not have `IsNaN` nodes, though `remove_isnan_in_attention_mask=True` could be nice to have, to be a bit more defensive for future models which might have `IsNaN`

File: scripts/e2e_eval/testsets/models_with_acc.json [hunk 9]
```diff
@@ -1431,5 +1550,100 @@
         "label_column": "fine_label"
       }
     }
+  },
```
Comment (vortex-captain) [inline scripts/e2e_eval/testsets/models_with_acc.json:1543]: seems that only new entries have `precision` field, maybe later we could add `precision` to all entries for clarity?

File: scripts/e2e_eval/cache/baseline_cache.json [hunk 10]
```diff
@@ -828,5 +828,55 @@
     },
     "elapsed": 222.8,
     "command": "python.exe run_pytorch_baseline.py --model fashion-clip --task zero-shot-image-classification --device cpu --num-samples 1000 --dataset fashion_mnist --split test --columns-mapping {\"input_column\": \"image\"}"
+  },
```
Comment (vortex-captain) [inline scripts/e2e_eval/cache/baseline_cache.json:831]: Now that precision is a new variable in evaluation, I guess we should associate with all metrics their respective model precision.

File: src/winml/modelkit/eval/metrics/text_similarity.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+
+logger = logging.getLogger(__name__)
+
+
+def _levenshtein(a: str, b: str) -> int:
```
Comment (vortex-captain) [inline src/winml/modelkit/eval/metrics/text_similarity.py:27]: nit: in case the distance computation is slow, we could consider using packages like https://pypi.org/project/python-Levenshtein/ to speed up (though I'm not sure if it would install on any machine as compilation is involved)

File: scripts/e2e_eval/testsets/models_with_acc.json (hunk 9 above)
Comment (zhenchaoni) [reply scripts/e2e_eval/testsets/models_with_acc.json:1543]: yes, we can do later.

File: scripts/e2e_eval/cache/baseline_cache.json (hunk 10 above)
Comment (zhenchaoni) [reply scripts/e2e_eval/cache/baseline_cache.json:831]: This will be done by our 8 EP runs. We will run all built-in models in all EP for all precision we support.

File: src/winml/modelkit/eval/metrics/text_similarity.py (hunk 11 above)
Comment (zhenchaoni) [reply src/winml/modelkit/eval/metrics/text_similarity.py:27]: fixed

File: src/winml/modelkit/models/hf/vision_encoder_decoder.py (hunk 8 above)
Comment (zhenchaoni) [reply src/winml/modelkit/models/hf/vision_encoder_decoder.py]: fixed, also fixed a typo for clamp constant

**Review by vortex-captain — APPROVED** [review]

**Review by vortex-captain — APPROVED** [review]

**Review by vortex-captain — APPROVED** [review]
